File: linkedList/328_odd_even_linked_list.py
class Solution:
    # O(n) time | O(1) space - where n is the length of the input list
    def oddEvenList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head: return head
        
        odd, even = head, head.next
        evenHead = even
        
        while even and even.next:
            odd.next = even.next
            odd = odd.next
            even.next = odd.next
            even = even.next
        
        odd.next = evenHead
        return head

File: linkedList/328_odd_even_linked_list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
// O(n) time | O(1) space - where n is the number of nodes
func oddEvenList(head *ListNode) *ListNode {
    if head == nil {
        return nil
    }
    odd := head
    even, evenHead := head.Next, head.Next

    for even != nil && even.Next != nil {
        odd.Next = even.Next
        even.Next = even.Next.Next
        odd = odd.Next
        even = even.Next
    }

    odd.Next = evenHead

    return head
}
